services/web/server/src/simcore_service_webserver/folders/_folders_api.py

```diff
@@ -1,341 +1,351 @@
+    app: web.Application,
+    user_id: UserID,
+    folder_id: FolderID,
+    product_name: ProductName,
+) -> FolderGet:
+    folder_db = await folders_db.get(
+        app, folder_id=folder_id, product_name=product_name
+    )
+
+    workspace_is_private = True
+    user_folder_access_rights = AccessRights(read=True, write=True, delete=True)
+    if folder_db.workspace_id:
+        user_workspace_access_rights = await check_user_workspace_access(
+            app,
+            user_id=user_id,
+            workspace_id=folder_db.workspace_id,
+            product_name=product_name,
+            permission="read",
+        )
+        workspace_is_private = False
+        user_folder_access_rights = user_workspace_access_rights.my_access_rights
+
+    folder_db = await folders_db.get_for_user_or_workspace(
+        app,
+        folder_id=folder_id,
+        product_name=product_name,
+        user_id=user_id if workspace_is_private else None,
+        workspace_id=folder_db.workspace_id,
+    )
+    return FolderGet(
+        folder_id=folder_db.folder_id,
+        parent_folder_id=folder_db.parent_folder_id,
+        name=folder_db.name,
+        created_at=folder_db.created,
+        modified_at=folder_db.modified,
+        trashed_at=folder_db.trashed_at,
+        owner=folder_db.created_by_gid,
+        workspace_id=folder_db.workspace_id,
+        my_access_rights=user_folder_access_rights,
+    )
+
+
+async def list_folders(
+    app: web.Application,
+    user_id: UserID,
+    product_name: ProductName,
+    folder_id: FolderID | None,
+    workspace_id: WorkspaceID | None,
+    trashed: bool | None,
+    offset: NonNegativeInt,
+    limit: int,
+    order_by: OrderBy,
+) -> FolderGetPage:
+    workspace_is_private = True
+    user_folder_access_rights = AccessRights(read=True, write=True, delete=True)
+
+    if workspace_id:
+        user_workspace_access_rights = await check_user_workspace_access(
+            app,
+            user_id=user_id,
+            workspace_id=workspace_id,
+            product_name=product_name,
+            permission="read",
+        )
+        workspace_is_private = False
+        user_folder_access_rights = user_workspace_access_rights.my_access_rights
+        _workspace_query = WorkspaceQuery(
+            workspace_scope=WorkspaceScope.SHARED, workspace_id=workspace_id
+        )
+    else:
+        _workspace_query = WorkspaceQuery(workspace_scope=WorkspaceScope.PRIVATE)
+
+    if folder_id:
+        # Check user access to folder
+        await folders_db.get_for_user_or_workspace(
+            app,
+            folder_id=folder_id,
+            product_name=product_name,
+            user_id=user_id if workspace_is_private else None,
+            workspace_id=workspace_id,
+        )
+        _folder_query = FolderQuery(
+            folder_scope=FolderScope.SPECIFIC, folder_id=folder_id
+        )
+    else:
+        _folder_query = FolderQuery(folder_scope=FolderScope.ROOT)

```

Comment:
OK, so this is the new part. You have two different scopes to list folders: Specific and root.   

but, how do you list full-depth? i.e. including all the folders in the hierarchy?

Can you provide a test with each of this scopes and depth?

Reply:
as discussed currently we support only ROOT, SPECIFIC and ALL (I am adding a test now)